aten/src/ATen/native/IndexingUtils.h

```diff
@@ -0,0 +1,135 @@
+#include <ATen/ExpandUtils.h>
+#include <ATen/native/TensorIterator.h>
+
+namespace at { namespace native {
+
+[[noreturn]]
+static void invalid_mask(const Tensor & self, int64_t idx, const Tensor & mask, int64_t maskIdx) {
+  std::stringstream ss;
+  ss << "The shape of the mask " << mask.sizes() << " at index " << maskIdx;
+  ss << " does not match the shape of the indexed tensor " << self.sizes();
+  ss << " at index " << idx;
+  AT_INDEX_ERROR(ss.str());
+}
+
+
+static std::vector<Tensor> expandTensors(const Tensor & self, TensorList indices) {
+  // Expands ByteTensor (masks) or BoolTensor (masks) into the equivalent indexing by LongTensors
+  std::vector<Tensor> result;
+  for (auto & index : indices) {
+    if (index.scalar_type() == kByte || index.scalar_type() == kBool) {
+      // The sizes of the ByteTensor mask or bool tensor must match the sizes of the
+      // corresponding dimensions in self
+      for (int64_t j = 0; j < index.dim(); j++) {
+        int64_t srcIdx = result.size() + j;
+        if (index.size(j) != self.size(srcIdx)) {
+          invalid_mask(self, srcIdx, index, j);
+        }
+      }
+      // Replace with nonzeros
+      auto nonzero = index.nonzero();
+      auto special_empty = false;
+      for (int64_t j = 0; j < index.dim(); j++) {
+        if (special_empty) {
+          // We can't call select on an empty tensor so we just create an empty
+          // tensor.
+          result.emplace_back(at::empty({0}, nonzero.options()));
+        } else {
+          result.emplace_back(nonzero.select(1, j));
+        }
+      }
+    } else {
+      result.emplace_back(index);
+    }
+  }
+  return result;
+}
+
+
+static void checkIndexTensorTypes(TensorList indices) {
+  for (auto& tensor : indices) {
+    if (tensor.defined()) {
+      auto scalarType = tensor.scalar_type();
+      if (scalarType != kLong && scalarType != kByte && scalarType != kBool) {
+          AT_INDEX_ERROR("tensors used as indices must be long, byte or bool tensors");
+      }
+    }
+  }
+}
+
+static bool hasContiguousSubspace(TensorList tl) {
```

Comment:
nit: It is a little confusing to use the word "contiguous" here since normally it refers to contiguity of strides (no action needed, just remarking)